### ISSUE TITLE
Family and Subset separation (a.k.a. hello 'Variants' )

### ISSUE THREAD

**Issue by mkolar** [getavalon/core#269]
Discussion
----------------

Having read the documentation that exists thus far and variety of issues brought up in the repo, I have to say that I find the concept of subset as it exists now slightly misleading and confusing. At first glance it makes sense storing all subsets directly within the publish directory a differentiating them by name (and their metadata of course) but I keep thinking that you might be missing a hierarchical level in the system that could simplify a lot for both api access, but also in terms of human readability.

It's also directly related to #246 (and I believe would inherently solve it)

Right now the subset name name seem fully arbitrary, albeit with proposed structure. So a subset of  `look`  family should be called `lookDefault`, `lookWet` etc., however what is stopping the artist from naming it just `default` or `wet`? At which point there's no way to tell what it actually contains without reading the metadata from inside.

Proposal
-----------

A solution that fixes this for both api access and direct inspection could be to split subset into 2 hierarchical levels: `family` and `variant` (could stay 'subset' of course I'm just using variant out of habit and no to confuse with current system)

This would make filtering easier, but most importantly it would unburden the artist from having to consider what he's creating when naming subsets. If he's chosen to publish a rig already, he shouldn't have to put it in the name and potentially introduce inconsistency if he does it wrong.

This is what the publish folder would look like
- **ASSET**
    - publish
        - family01
            - variant01
                - versions...
            - variant02

Example:
- **BOB**
  - publish
    - _model_
      - default
        - v01
        - v02
    - _look_
      - default
      - wet

I fully realise that this project might just be a tad too far for such type of change , but I believe it could be implemented with backwards compatibility as an alternative. Studio could choose to use subsets of family/variant workflow.

On a side note. I can't help but think, that subset is a term that needs to be explained to an artist, whereas when you tell someone that this character has 3 different look variants, it's understandable without extra explanation.

**Comment by mottosso** [comment]
Thanks @mkolar! Having a think.

**Comment by mottosso** [comment]
Ok, I spot three topics.

1. Vocabulary, "Variant" conveys more meaning than "Subset"
2. Directory Layout, whether it can be made more clear
3. Directory Layout, Avalon versus use of Avalon

<br>

> At first glance it makes sense storing all subsets directly within the publish directory a differentiating them by name

Before diving in, Avalon does not involve directory structure and I'll try to highlight how.

The directory structure you are referring to is the one Mindbender is using and is provided by the default project template as an example.

It looks like this.

```python
{root}/{project}/{silo}/{asset}/publish/{subset}/v{version:0>3}/{subset}.{representation}
```

- [Reference](https://getavalon.github.io/2.0/reference/#project-configuration-api)

It comes from (1) a "path template" stored as metadata stored per project and (2) a Pyblish plug-in applying a dictionary of values to this template.

You could achieve what you are looking for like this.

```python
{root}/{project}/{silo}/{asset}/publish/{subset}/{family}/v{version:0>3}/{subset}.{representation}
```

You can implement this by for example using Mindbender's integrator and adding `family` to the dictionary of keys currently available to where paths are used. At Mindbender, this path is generated after extraction (from e.g. Maya) and during integration.

```python
data = {
    "root": api.registered_root(),
    "project": PROJECT,
    "silo": SILO,
    "asset": ASSET,
    "family": version["data"]["family"],
    "subset": subset["name"],
    "version": version["name"],
}

path = (
  "{root}/{project}/{silo}/{asset}/publish/"
  "{subset}/{family}/v{version:0>3}/"
  "{subset}.{representation}"
).format(**data)
```

- [Reference](https://github.com/mindbender-studio/config/blob/49e32c66cc1fa7cf86540a2f61fc6f8b85646ef2/polly/plugins/publish/integrate_asset.py#L142)

The idea behind it is to utilise Python's `.format` mechanism that takes a series of key/value pairs to fill in a string of "slots", such as `{subset}`. And again, the storage and use of this data is entirely up to the user of Avalon and nothing Avalon depends on. For example, you could choose to store a path template per asset as a Lucidity string or absolute paths of some kind, or paths relative various global locations. Or not store them as metadata at all.

### Avalon vs Use of Avalon

At this point I'd expect you to ask "but then what does Avalon do?" which is a fair question. It's the equivalent of Qt and using Qt. In this case, Avalon provides you the ability to store and use a path template in production, the fact that there is a document-database backend and consistent object model for all data in every department of every production.

This particular issue is a question of workflow and studio implementation, like how the implementation of a Qt window is constructed and whether it's using the Signal and Slots mechanism properly. In this scenario, Avalon is closer to Qt than it is to that particular version of a window, where the window is [Mindbender's use of Avalon](https://github.com/mindbender-studio/config).

We should be able to talk about both of these things, like talking about Qt itself and a particular application written in Qt. Both have best practices and lots to discuss. I'm not yet entirely sure how to separate between the discussions, both technically such as in two repositories, but mainly socially. How to communicate what is Qt and what is a program using Qt.

Ideally this project - core - should only involve core issues and use of it should reside in the corresponding repository. But it can understandably be difficult to know when to separate anything without being an expert at either.

Do continue having these discussions here though until we figure that out. I'd imagine it's a matter of scale and right now there is only one prominent example of using Avalon, which is Mindbender and Colorbleed which are mostly similar.

### Vocabulary and Variant

Personally, either is fine to me. I'd be happy to consider a PR whereby either is an option to the user. Ultimately, I'd say it's fine to use one of these for code and another in speak, whereby the spoken word is the interface to the underlying software. Software can change and studio vocabulary should be allowed to grow independently of it.

**Comment by mkolar** [comment]
Thanks for great clarification.

> At this point I'd expect you to ask "but then what does Avalon do?" which is a fair question. 

Actually, it makes perfect sense. On the contrary, from looking at the project so far I assumed it was trying to take on more built in responsibility than would be healthy, without leaving much breathing space for customisation and different workflows. Didn't realise the depth to which the Project configuration can go. So thumbs up from me.

Looks like it should be fairly straightforward to transition from our current system (using lucidity with yaml templates) to avalon on some smaller test project.

> Ultimately, I'd say it's fine to use one of these for code and another in speak,

Absolutely. As far as the functionality to define the structures is there. It's a nuisance.

**Comment by mottosso** [comment]
> Looks like it should be fairly straightforward to transition from our current system (using lucidity with yaml templates) to avalon on some smaller test project.

I'm looking to put together a "Learn By Example" style tutorial (much like http://learn.pyblish.com) on how to do just that. So as you go, I would be most interested in spectating and gathering feedback on how it goes with the goal of simplifying the install and initial configuration as much as possible.

Let me know when you get going, here is the best starting point currently.

- https://getavalon.github.io/2.0/guides/#install

**Comment by BigRoy** [comment]
@mkolar I'm with you on this. I'm expecting a Subset (== "Variant" to me) to be of a static family type too. So them being inside a "family" folder would also be expected on my end as default.

Note that this is also somewhat related to: https://github.com/getavalon/core/issues/246

**Comment by mkolar** [comment]
> Note that this is also somewhat related to: #246

I'd say mery much directly related :)

**Comment by BigRoy** [comment]
With #246 implemented by #443 I think this discussion can come to and end. The family is **not** separated from the subset, however the subset does get locked to a specific family (so it's even *more* linked!)

As opposed to having multiple subsets `default` in an asset (e.g. `hero`) of different families (`model`, `look`) the naming convention for subsets is to include the family to `lookDefault` and `modelDefault`. (To ensure they are unique) 

As such I believe through this discussion the outcome of separating family and subset is that we are *not* doing so in the database. You're still free to design it inside your published filepaths through configuring the templates for the project.

With that I think we can **Close** this issue. Feel free to reopen @mkolar @mottosso if you feel whether continuing this discussion is still required.

**Comment by mkolar** [comment]
I agree that this is working as it should now.
